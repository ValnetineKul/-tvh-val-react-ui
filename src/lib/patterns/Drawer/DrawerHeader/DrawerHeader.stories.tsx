import { action } from '@storybook/addon-actions';
import DrawerHeader from './DrawerHeader';
import { createMeta, createTemplate } from '../../../story-utils';

const Template = createTemplate(DrawerHeader);

export const Primitive = Template.bind({});
Primitive.args = {
  isLogo: true,
  onClose: action('onClose'),
};
Primitive.argTypes = {
  isLogo: {
    control: {
      disable: true,
    },
  },
  component: {
    control: {
      disable: true,
    },
  },
  heading: {
    table: {
      disable: true,
    },
  },
  onStartButtonClick: {
    table: {
      disable: true,
    },
  },
};
Primitive.storyName = 'Logo';

export const Heading = Template.bind({});
Heading.args = {
  heading: 'Main navigation',
  onClose: action('onClose'),
};
Heading.argTypes = {
  isLogo: {
    table: {
      disable: true,
    },
  },
  onStartButtonClick: {
    table: {
      disable: true,
    },
  },
  component: {
    table: {
      disable: true,
    },
  },
  href: {
    table: {
      disable: true,
    },
  },
  to: {
    table: {
      disable: true,
    },
  },
  target: {
    table: {
      disable: true,
    },
  },
};

export const HeadingWithStartButton = Template.bind({});
HeadingWithStartButton.args = {
  heading: 'Main navigation',
  onStartButtonClick: action('onStartButtonClick'),
  onClose: action('onClose'),
};
HeadingWithStartButton.argTypes = {
  isLogo: {
    table: {
      disable: true,
    },
  },
  component: {
    table: {
      disable: true,
    },
  },
  href: {
    table: {
      disable: true,
    },
  },
  to: {
    table: {
      disable: true,
    },
  },
  target: {
    table: {
      disable: true,
    },
  },
};
HeadingWithStartButton.storyName = 'HeadingWithStartButton';

export default createMeta({
  component: DrawerHeader,
  title: 'Patterns/Drawers/DrawerHeader',
  argTypes: {
    disabled: {
      table: {
        disable: true,
      },
    },
    download: {
      table: {
        disable: true,
      },
    },
  },
});
