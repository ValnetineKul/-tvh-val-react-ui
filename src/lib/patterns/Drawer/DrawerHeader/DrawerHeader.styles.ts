import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles({ name: 'DrawerHeader' })((theme) => ({
  root: {},
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    minHeight: 48,
    paddingLeft: theme.layoutSpacing['spacing/400'],
    paddingRight: theme.layoutSpacing['spacing/400'],
    position: 'sticky',
    top: 0,
    left: 0,
    zIndex: 1,
  },
  headingWrapper: {
    display: 'flex',
    alignItems: 'center',
    width: 'calc(100% - 32px)',
  },
  heading: {
    marginLeft: theme.layoutSpacing['spacing/300'],
  },
  truncatedHeading: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  },
}));

export default useStyles;
