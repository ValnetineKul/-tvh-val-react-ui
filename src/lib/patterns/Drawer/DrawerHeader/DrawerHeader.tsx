import React, { useEffect, useRef, useState } from 'react';
import type { FC } from 'react';
import { AngleLeft, Times } from '../../../components/Icon/icons/functional';
import Icon from '../../../components/Icon';
import IconButton from '../../../components/Buttons/IconButton';
import Logo from '../../../components/Logo';
import Surface from '../../../components/Surfaces/Surface';
import Tooltip from '../../../components/Tooltip';
import Typography from '../../../components/Typography';
import type { ButtonBaseProps } from '../../../components/ButtonBase';
import type { DataAttributes } from '../../../types/common';
import { useButtonBase } from '../../../components/ButtonBase';
import useStyles from './DrawerHeader.styles';

interface CommonProps extends ButtonBaseProps {
  onClose: () => void;
  className?: string;
  closeButtonProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
}

interface LogoProps {
  isLogo: boolean;
  heading?: never;
  onStartButtonClick?: never;
}

interface HeadingProps {
  heading: string;
  onStartButtonClick?: () => void;
  isLogo?: never;
}

export type DrawerHeaderProps = CommonProps & (LogoProps | HeadingProps);

const DrawerHeader: FC<DrawerHeaderProps> = ({ children, onClose, className, closeButtonProps, ...props }) => {
  const { classes, cx } = useStyles();

  const isLogo = 'isLogo' in props && props.isLogo;
  const heading = 'heading' in props && props.heading;
  const onStartButtonClick = 'onStartButtonClick' in props ? props.onStartButtonClick : undefined;

  const isShowLogo = isLogo || !heading;
  const buttonBaseProps = useButtonBase(props);

  const headingRef = useRef<HTMLElement>(null);
  const [shouldShowTooltip, setShouldShowTooltip] = useState(false);

  useEffect(() => {
    const compareSize = () => {
      setShouldShowTooltip(
        !!(
          heading &&
          headingRef &&
          headingRef.current &&
          headingRef.current.scrollWidth > headingRef.current.clientWidth
        )
      );
    };

    compareSize();

    window.addEventListener('resize', compareSize);
    return () => {
      window.removeEventListener('resize', compareSize);
    };
  }, [heading]);

  return (
    <Surface color="Primary">
      <div className={cx(classes.header, className)}>
        {isShowLogo && <Logo variant="negative" size="sm" {...buttonBaseProps} />}
        {heading && (
          <div className={classes.headingWrapper}>
            {onStartButtonClick && (
              <IconButton aria-label="back" size="md" onClick={onStartButtonClick} icon={<Icon icon={AngleLeft} />} />
            )}
            <Tooltip label={heading} disableHoverListener={!shouldShowTooltip} className={classes.truncatedHeading}>
              <Typography
                variant="h4"
                ref={headingRef}
                className={cx(classes.truncatedHeading, { [classes.heading]: !!onStartButtonClick })}
              >
                {heading}
              </Typography>
            </Tooltip>
          </div>
        )}
        <IconButton {...closeButtonProps} aria-label="close" size="md" onClick={onClose} icon={<Icon icon={Times} />} />
      </div>
    </Surface>
  );
};

export default DrawerHeader;
