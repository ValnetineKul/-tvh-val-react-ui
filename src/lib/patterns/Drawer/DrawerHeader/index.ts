import DrawerHeader from './DrawerHeader';

export type { DrawerHeaderProps } from './DrawerHeader';
export default DrawerHeader;
