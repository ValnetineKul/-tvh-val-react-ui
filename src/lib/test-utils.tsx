import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
