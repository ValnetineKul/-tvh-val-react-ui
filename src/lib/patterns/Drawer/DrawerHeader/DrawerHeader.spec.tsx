import React from 'react';
import type { ComponentProps } from 'react';
import { render, screen } from '../../../test-utils';
import DrawerHeader from './DrawerHeader';
import { Primitive, Heading, HeadingWithStartButton } from './DrawerHeader.stories';
import type { Story } from '../../../story-utils';

describe('Drawer', () => {
  describe('Should render stories correctly', () => {
    const cases: [string, Story<ComponentProps<typeof DrawerHeader>>][] = [
      ['logo', Primitive],
      ['heading', Heading],
      ['heading with start button', HeadingWithStartButton],
    ];
    test.each(cases)('story: %p', (testName, storybookCase) => {
      const { container } = render(
        <DrawerHeader {...storybookCase.args}>
          <div>Content</div>
        </DrawerHeader>
      );
      expect(container.firstChild).toMatchSnapshot(testName);
    });
  });

  it('Should render heading', () => {
    const onClose = jest.fn();
    render(
      <DrawerHeader onClose={onClose} heading="Test">
        <div>Content</div>
      </DrawerHeader>
    );
    expect(screen.getByText(/Test/)).toBeInTheDocument();
    expect(screen.queryByAltText(/logo/)).not.toBeInTheDocument();
  });

  it('Should render heading with start button', () => {
    const onClose = jest.fn();
    const onStartButtonClick = jest.fn();
    render(
      <DrawerHeader onClose={onClose} heading="Test" onStartButtonClick={onStartButtonClick}>
        <div>Content</div>
      </DrawerHeader>
    );
    expect(screen.getByText(/Test/)).toBeInTheDocument();
    expect(screen.getByLabelText(/back/, { selector: 'button' })).toBeInTheDocument();
    expect(screen.queryByAltText(/logo/)).not.toBeInTheDocument();
  });

  it('Should render logo if heading is empty', () => {
    const onClose = jest.fn();
    render(
      <DrawerHeader onClose={onClose} heading="">
        <div>Content</div>
      </DrawerHeader>
    );
    expect(screen.getByAltText(/Logo/)).toBeInTheDocument();
  });
});
