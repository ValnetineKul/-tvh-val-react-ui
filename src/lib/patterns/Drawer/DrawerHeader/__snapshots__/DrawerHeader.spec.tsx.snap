// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Should render stories correctly story: "heading with start button": heading with start button 1`] = `
<div
  class="tss-z4jt7l-Surface-root"
>
  <div
    class="tss-zcjtpr-DrawerHeader-header"
  >
    <div
      class="tss-1magt3-DrawerHeader-headingWrapper"
    >
      <button
        aria-label="back"
        class="MuiButtonBase-root tss-7kteag-ButtonBase-root-ButtonBase-focusOut-IconButton-root mui-10d1a0h-MuiButtonBase-root"
        tabindex="0"
        type="button"
      >
        <div
          class="tss-w1sjks-Icon-root-Icon-sizeMd"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
            color="inherit"
            focusable="false"
            viewBox="0 0 24 24"
          >
            angleLeft.svg
          </svg>
        </div>
      </button>
      <div
        aria-label="Main navigation"
        class="tss-3-Typography-weightRegular-ref tss-1jbpm2b-Typography-root-Typography-weightRegular-Tooltip-wrapper-DrawerHeader-truncatedHeading-Typography-body400"
        data-mui-internal-clone-element="true"
      >
        <h4
          class="tss-3-Typography-headerFunctional-ref tss-1vfq91l-Typography-root-Typography-headerFunctional-DrawerHeader-truncatedHeading-DrawerHeader-heading-Typography-h4"
        >
          Main navigation
        </h4>
      </div>
    </div>
    <button
      aria-label="close"
      class="MuiButtonBase-root tss-7kteag-ButtonBase-root-ButtonBase-focusOut-IconButton-root mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <div
        class="tss-w1sjks-Icon-root-Icon-sizeMd"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
          color="inherit"
          focusable="false"
          viewBox="0 0 24 24"
        >
          times.svg
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`Drawer Should render stories correctly story: "heading": heading 1`] = `
<div
  class="tss-z4jt7l-Surface-root"
>
  <div
    class="tss-zcjtpr-DrawerHeader-header"
  >
    <div
      class="tss-1magt3-DrawerHeader-headingWrapper"
    >
      <div
        aria-label="Main navigation"
        class="tss-3-Typography-weightRegular-ref tss-1jbpm2b-Typography-root-Typography-weightRegular-Tooltip-wrapper-DrawerHeader-truncatedHeading-Typography-body400"
        data-mui-internal-clone-element="true"
      >
        <h4
          class="tss-3-Typography-headerFunctional-ref tss-1r52m5v-Typography-root-Typography-headerFunctional-DrawerHeader-truncatedHeading-Typography-h4"
        >
          Main navigation
        </h4>
      </div>
    </div>
    <button
      aria-label="close"
      class="MuiButtonBase-root tss-7kteag-ButtonBase-root-ButtonBase-focusOut-IconButton-root mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <div
        class="tss-w1sjks-Icon-root-Icon-sizeMd"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
          color="inherit"
          focusable="false"
          viewBox="0 0 24 24"
        >
          times.svg
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`Drawer Should render stories correctly story: "logo": logo 1`] = `
<div
  class="tss-z4jt7l-Surface-root"
>
  <div
    class="tss-zcjtpr-DrawerHeader-header"
  >
    <button
      class="MuiButtonBase-root tss-x4btvw-ButtonBase-root-ButtonBase-focusOut-Logo-root mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <img
        alt="Logo"
        class="tss-iftyg6-Image-root-Logo-sizeSm"
        src="negativeLogo.svg"
      />
    </button>
    <button
      aria-label="close"
      class="MuiButtonBase-root tss-7kteag-ButtonBase-root-ButtonBase-focusOut-IconButton-root mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <div
        class="tss-w1sjks-Icon-root-Icon-sizeMd"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
          color="inherit"
          focusable="false"
          viewBox="0 0 24 24"
        >
          times.svg
        </svg>
      </div>
    </button>
  </div>
</div>
`;
